src/blockchain/ethereum.rs

```diff
@@ -41,119 +46,234 @@ impl Ethereum {
     ///
     /// * `address` - The address of an ethereum node.
     /// * `validator` - The address of the validator to sign and send messages from.
-    pub fn new(address: &str, validator: Address) -> Result<Self, Error> {
-        let (event_loop, http) = match Http::new(address) {
-            Ok((event_loop, http)) => (event_loop, http),
-            Err(error) => {
-                error!("Could not connect to ethereum: {}", error);
-                return Err(Error::new(ErrorKind::NodeError, error.to_string()));
-            }
-        };
+    pub fn new(
+        address: &str,
+        validator: Address,
+        event_loop: Box<tokio_core::reactor::Handle>,
+    ) -> Self {
+        let http = Http::with_event_loop(address, &event_loop, 5)
+            .expect("Could not initialize ethereum HTTP connection");
         let web3 = Web3::new(http);
 
         let password = rpassword::prompt_password_stdout(&format!(
             "Please enter the password for account {:x}: ",
             &validator,
         )).unwrap();
 
-        let ethereum = Ethereum {
+        Ethereum {
             web3,
             validator,
             password,
-            _event_loop: event_loop,
-        };
+            event_loop,
+        }
+    }
 
-        ethereum.unlock_account();
+    /// Stream blocks returns a `futures::stream::Stream` of `Block`s.
+    ///
+    /// Converts a stream of web3 blocks to a stream of blocks.
+    ///
+    /// It is the caller's responsibility to poll the stream, e.g. call `for_each` and put the
+    /// future into a reactor.
+    pub fn stream_blocks(&self) -> impl Stream<Item = Block, Error = Error> {
+        // Blocks filter is a future that returns a filter.
+        let blocks_filter = self.web3.eth_filter().create_blocks_filter();
+
+        // Block hashes is a stream of block hashes.
+        let block_hashes = blocks_filter
+            .map(|filter| filter.stream(Duration::from_secs(1)))
+            .flatten_stream();
+
+        // Web3 blocks is a stream of block futures, mapped from a stream of block hashes.
+        let web3_clone = self.web3.clone();
```

Comment:
Because of this:
```rust
            and_then(move |block_hash| {
                web3_clone
                    .eth()
                    .block(BlockId::from(block_hash))
                    .map_err(|error| {
                        Error::new(
                            ErrorKind::NodeError,
                            format!("Was not able to retrieve block: {}", error),
                        )
                    })
            });
```

This closure accesses `web3_clone`. If the closure would access `self.web3` instead, the compiler couldn't guarantee that `self` still exists. `&self` is passed into `stream_blocks()` by reference and the `stream_blocks()` function returns before the closure is run (in fact, it's a future in this case and may run much later). The compiler doesn't know, whether `self`, which was only borrowed (`&self`), would still be available when the closure is run.
The clone, on the other hand, is not borrowed. And it is moved into the closure by the `move` keyword.